fdbserver/DeltaTree.h

```diff
@@ -899,3 +899,738 @@ struct DeltaTree {
+
+			int nIndex = rootIndex();
+			int parentIndex = nIndex;
+			DecodedNode* parentDecoded;
+			// Result of comparing node at parentIndex
+			int cmp = 0;
+			// Height of the inserted node
+			int height = 0;
+
+			// Find the parent to add the node to
+			// This is just seek but modifies parentIndex instead of nodeIndex and tracks the insertion height
+			deltatree_printf(
+			    "insert(%s) start %s\n", k.toString().c_str(), Cursor(cache, tree, parentIndex).toString().c_str());
+			while (nIndex != -1) {
+				++height;
+				parentIndex = nIndex;
+				parentDecoded = &cache->get(parentIndex);
+				cmp = k.compare(get(*parentDecoded), skipLen);
+				deltatree_printf("insert(%s) moved cmp=%d %s\n",
+				                 k.toString().c_str(),
+				                 cmp,
+				                 Cursor(cache, tree, parentIndex).toString().c_str());
+
+				if (cmp == 0) {
+					break;
+				}
+
+				if (cmp > 0) {
+					deltatree_printf("insert(%s) move right\n", k.toString().c_str());
+					nIndex = getRightChildIndex(nIndex);
+				} else {
+					deltatree_printf("insert(%s) move left\n", k.toString().c_str());
+					nIndex = getLeftChildIndex(nIndex);
+				}
+			}
+
+			// If the item is found, mark it erased if it isn't already
+			if (cmp == 0) {
+				DeltaT& delta = tree->nodeAt(parentDecoded->nodeOffset)->delta(tree->largeNodes);
+				if (delta.getDeleted()) {
+					delta.setDeleted(false);
+					++tree->numItems;
+					tree->nodeBytesDeleted -= (delta.size() + Node::headerSize(tree->largeNodes));
+					deltatree_printf("insert(%s) deleted item restored %s\n",
+					                 k.toString().c_str(),
+					                 Cursor(cache, tree, parentIndex).toString().c_str());
+					return true;
+				}
+				deltatree_printf("insert(%s) item exists %s\n",
+				                 k.toString().c_str(),
+				                 Cursor(cache, tree, parentIndex).toString().c_str());
+				return false;
+			}
+
+			// If the tree was empty or the max insertion height is exceeded then fail
+			if (parentIndex == -1 || height > maxHeightAllowed) {
+				return false;
+			}
+
+			// Find the base base to borrow from, see if the resulting delta fits into the tree
+			int leftBaseIndex, rightBaseIndex;
+			bool addingRight = cmp > 0;
+			if (addingRight) {
+				leftBaseIndex = parentIndex;
+				rightBaseIndex = parentDecoded->rightParentIndex;
+			} else {
+				leftBaseIndex = parentDecoded->leftParentIndex;
+				rightBaseIndex = parentIndex;
+			}
+
+			T leftBase = leftBaseIndex == -1 ? cache->lowerBound : get(cache->get(leftBaseIndex));
+			T rightBase = rightBaseIndex == -1 ? cache->upperBound : get(cache->get(rightBaseIndex));
+
+			int common = leftBase.getCommonPrefixLen(rightBase, skipLen);
+			int commonWithLeftParent = k.getCommonPrefixLen(leftBase, common);
+			int commonWithRightParent = k.getCommonPrefixLen(rightBase, common);
+			bool borrowFromLeft = commonWithLeftParent >= commonWithRightParent;
```

Comment:
Maybe, and if it's simple we should do it, but I don't know that we will notice a difference without a very specific workload.

The call chain length should almost always be 1 since the left and right parent indices are known and so can be accessed with equal indirection.

For seeks to an uncached node, a comparison is done to every node along the traversal path, so all ancestor keys are already cached.  

For cursor movements to uncached adjacent nodes, there is the possibility of traversing multiple left-only or multiple right-only links, passing and uncached nodes and not reading or caching them.  Once the final node is reached, it would be decompressed, recursively decompressing _and caching_ any base ancestors along the way in a single pass (though the second pass after the downward traversal pass) until it reaches one that is already decoded and cached.   In the event a cursor movement follows a left/right-only downward path to an uncached node with equal borrowing potential from either side, it is less work if it would borrow from the right/left ancestor, respectively.  

However, such cursor movement would be part of either a read range scan or the commit path which also iterates, so it is likely that some of the additional decoded nodes are about to be needed anyway.

I wonder if there would be some benefit from memory locality.  For a seek, all ancestor keys would be in the CPU cache, but when a cursor is moved intermittently it might be beneficial to prefer ancestors of equal prefix length that are on the side from which the cursor is moving.  Notably, the bias for base selection requires knowing the cursor movement direction and is not solely based on how far up the ancestry each possible base node is.  It's probably safe, however, to assume that the vast majority of iteration is in the forward direction.

All that said, this might be easy to implement so I'd like to try it, but not right now.